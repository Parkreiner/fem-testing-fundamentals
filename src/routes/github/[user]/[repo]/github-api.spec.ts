import { describe, it, vi } from "vitest";
import { Fetch, GithubApi } from "./github-api";

function getMockPromise<T = unknown>(): Promise<T> & {
  resolve: (value: T) => void;
  reject: (error: any) => void;
} {
  let resolve!: (value: T) => void;
  let reject!: (error: unknown) => void;

  const promise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  }) as any;

  promise.resolve = resolve;
  promise.reject = reject;
  return promise;
}

describe("github-api", () => {
  describe("getRepository", () => {
    it("should return repository information", async ({ expect }) => {
      const fetch = vi.fn<Parameters<Fetch>, ReturnType<Fetch>>(getMockPromise);

      const api = new GithubApi("TOKEN", fetch);
      const responsePromise = api.getRepository("USERNAME", "REPOSITORY");

      expect(fetch).toHaveBeenCalledWith(
        "https://api.github.com/repos/USERNAME/REPOSITORY",
        {
          headers: {
            "User-Agent": "Qwik Workshop",
            "X-GitHub-Api-Version": "2022-11-28",
            Authorization: "Bearer TOKEN",
          },
        },
      );

      fetch.mock.results[0].value.resolve(new Response('"BLAH"'));
      expect(await responsePromise).toEqual("BLAH");
    });

    it.skip("should timeout after x seconds with time out response");
  });
});
