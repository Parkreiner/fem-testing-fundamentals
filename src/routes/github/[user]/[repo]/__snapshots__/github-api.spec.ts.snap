// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`github-api > getRepository > should return repository information 1`] = `
{
  "documentation_url": "https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting",
  "message": "API rate limit exceeded for 204.16.241.141. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)",
}
`;
